extension/src/app/cloudwatch-widget.ts

```diff
@@ -1,20 +1,79 @@
-import * as SDK from 'azure-devops-extension-sdk';
+import { notifyLoadSucceeded } from 'azure-devops-extension-sdk';
+import { ServiceEndpointRestClient } from 'azure-devops-extension-api/ServiceEndpoint';
+import { getClient } from 'azure-devops-extension-api';
+import { getProjectId, getSideLength } from './utils';
 
 export class CloudWatchWidget {
     private widget: HTMLDivElement;
+    private image: HTMLImageElement;
+    private projectId: string;
+    private endpointId: string;
+    private widgetJson: any;
+    private height: number;
+    private width: number;
+
+    constructor(
+        private endpointClient: ServiceEndpointRestClient = getClient(ServiceEndpointRestClient)
+    ) { }
 
     public async preload() {
+        this.projectId = await getProjectId();
         return {};
     }
 
     public async load(settings) {
+        this.setDimensions(settings.size);
         this.render(settings.customSettings.data);
         return {};
     }
 
-    private render(query: string) {
-        this.widget = document.querySelector('.widget');
-        this.widget.innerText = query;
-        SDK.notifyLoadSucceeded();
+    private async render(data: string) {
+        this.setSettings(data);
+        this.setupElements();
+        this.image.src = await this.retrieveImageSrc();
+        notifyLoadSucceeded();
+    }
+
+    private setupElements(): void {
+        if (!this.widget) {
+            this.widget = document.querySelector('.widget');
+        }
+        if (!this.image) {
+            this.image = document.createElement('img');
+            this.widget.appendChild(this.image);
+        }
+    }
+
+    private setDimensions(size: {columnSpan: number, rowSpan: number }): void {
```

Comment:
Worth leveraging destructuring on the param here IMO. Beyond the more idiomatic/stylistic aspects with ES6+, it also prevents the potential for runtime TypeErrors on the key access should `size` be some flavor of falsy.

I'm sure you're not looking for this type of feedback at this point in the dev cycle, but food for thought :smile:

Reply:
makes sense, a good consideration for cleanup

Reply:
done

Reply:
i think this can be expressed directly within the function signature and save the need for current line 48, but probably not worth fighting tsc if it barks (I know this tact is possible but my mental tsc is not as good as the real one and i may be misremembering precise syntax)

```suggestion
    private setDimensions({columnSpan: number, rowSpan: number }): void {
```

Reply:
oh thats cool

Reply:
no done lol